File: tests/scala/src/main/scala/sertests/MemoryFormatWriter.scala
package sertests

import zio.{IO, UIO, Chunk}
import dev.argon.verilization.scala_runtime.FormatWriter
import java.io.ByteArrayOutputStream

final class MemoryFormatWriter private(stream: ByteArrayOutputStream) extends FormatWriter[Any, Nothing] {
    override def writeByte(b: Byte): UIO[Unit] =
        IO.effectTotal { stream.write(b) }

    override def writeShort(s: Short): UIO[Unit] =
        IO.effectTotal {
            stream.write(s.toByte)
            stream.write((s >>> 8).toByte)
        }

    override def writeInt(i: Int): UIO[Unit] =
        IO.effectTotal {
            stream.write(i.toByte)
            stream.write((i >>> 8).toByte)
            stream.write((i >>> 16).toByte)
            stream.write((i >>> 24).toByte)
        }

    override def writeLong(l: Long): UIO[Unit] =
        IO.effectTotal {
            stream.write(l.toByte)
            stream.write((l >>> 8).toByte)
            stream.write((l >>> 16).toByte)
            stream.write((l >>> 24).toByte)
            stream.write((l >>> 32).toByte)
            stream.write((l >>> 40).toByte)
            stream.write((l >>> 48).toByte)
            stream.write((l >>> 56).toByte)
        }

    override def writeBytes(data: Chunk[Byte]): UIO[Unit] =
        IO.effectTotal {
            stream.write(data.toArray)
        }

    def toChunk: UIO[Chunk[Byte]] =
        IO.effectTotal { Chunk.fromArray(stream.toByteArray()) }
}

object MemoryFormatWriter {
    def make: UIO[MemoryFormatWriter] =
        IO.effectTotal {
            new MemoryFormatWriter(new ByteArrayOutputStream())
        }
}

File: tests/java/src/main/java/sertests/MemoryFormatWriter.java
package sertests;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

import dev.argon.verilization.runtime.FormatWriter;

public final class MemoryFormatWriter implements FormatWriter {

    private final ByteArrayOutputStream stream = new ByteArrayOutputStream();

    @Override
	public void writeByte(byte b) throws IOException {
        stream.write(b);
    }

    @Override
	public void writeShort(short s) throws IOException {
        writeByte((byte)s);
        writeByte((byte)(s >>> 8));
    }

    @Override
	public void writeInt(int i) throws IOException {
        writeShort((short)i);
        writeShort((short)(i >>> 16));
    }

    @Override
	public void writeLong(long l) throws IOException {
        writeInt((int)l);
        writeInt((int)(l >>> 32));
    }

    @Override
	public void writeBytes(byte[] data) throws IOException {
        stream.write(data);
    }

    public byte[] toByteArray() {
        return stream.toByteArray();
    }

}
